### ISSUE TITLE
ardb crashes during save or bgsave

### ISSUE THREAD

**Issue by harutyundermenjyan** [yinqiwen/ardb#246]
I have imported 1GB redis data into ardb/rocksdb. When I try to save or bgsave it crashes after 14MB.
Is it known issue or do you have any solution for it?

**Comment by yinqiwen** [comment]
Can u use gdb attach the core dump file to get the crash tack trace?

**Comment by harutyundermenjyan** [comment]
when I am trying to use gdb it hangs up.
1. ardb-server /etc/ardb.conf
2. gdb ardb-server PID-of-the-above-command
do my steps right?
P.S.
the last line of gdb output is:
81	../sysdeps/unix/syscall-template.S: No such file or directory.

**Comment by yinqiwen** [comment]
use gdb ./ardb-server <core file> can attach exe to a core file. But first u must exec ulimit -c unlimited before starting ardb-server to let it generate a core file after crash.  

and when u enter the gdb shell  
use `bt` to see current thread's stack trace
use `info threads` to list all threads
use `thread <id>` to switch thread

**Comment by harutyundermenjyan** [comment]
but is it ok to start without config file?
as I understand from your example above it will run without config file.

**Comment by yinqiwen** [comment]
`gdb ./ardb-server <core file>` would not start ardb server. We use it to debug the core dump file with exe file.  This would be done after the core file generated.   

First time, u should exec `ulimit -c unlimited` to enable core file generated after cash, then start the server normally to let it crash by `bgsave`

**Comment by yinqiwen** [comment]
I just recreated this issue and fixed it.

**Comment by harutyundermenjyan** [comment]
Thank you vary much. 
I was going to attach my core dump file, it just finished the download process from server. 
do I attach it or is no longer needed?

**Comment by yinqiwen** [comment]
No need to do that, it's fixed by the latest commit.

**Comment by harutyundermenjyan** [comment]
OK. Thanks again.
I will try it.

**Comment by harutyundermenjyan** [comment]
Dear yinqiwen 
we have an issue yet.
Now it is doesn't crashes but hangs up after 14MB and It indefinitely remains in save mode.

**Comment by harutyundermenjyan** [comment]
Also I have tried a different DBs as well. Does't help, the situation is the same.

**Comment by harutyundermenjyan** [comment]
Dear ardb Maintainer,
We have an issue yet.
Now it doesn't crashes but hangs up after ~14MB and It indefinitely remains
in save mode.
Anyway the save process stops when backup file size becomes ~14MB in my
case.

I have tried different redis dump files as well. It doesn't help, the
situation is the same.
-- 
Best regards,
Harutyun Dermenjyan

On Wed, May 25, 2016 at 6:31 PM, yinqiwen <notifications@github.com> wrote:

> No need to do that, it's fixed by the latest commit.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/yinqiwen/ardb/issues/246#issuecomment-221594689>
>



-- 
Best regards,
Harutyun Dermenjyan

**Comment by yinqiwen** [comment]
Any log in output?

**Comment by yinqiwen** [comment]
It's a bug happened when saving sorted set value into redis format rdb. Fixed by latest commit.

**Comment by harutyundermenjyan** [comment]
OK, seems now it perform a complete save/backup but a small issue still exists that is, it doesn't exiting from save mode after save complete.

**Comment by yinqiwen** [comment]
what's the 'save mode' mean? Is the redis-client never receive 'OK' from server?

**Comment by harutyundermenjyan** [comment]
I'm connecting to ardb trough redis-cli -p 16379, then import /tmp/dump.rdb and after the import completion just typing save, seem it saves normally and doesn't exit from save mode but in log file it prints that the save has been completed.

127.0.0.1:16379> save
``
`root@supermicro1:/var/lib/ardb# tail -f log/ardb-server.log
[25282] 05-26 10:05:09,455 DEBUG Process recved cmd cost 1028098287us
[25282] 05-26 10:10:00,132 INFO ========================Period Statistics Dump Begin===========================
[25282] 05-26 10:10:00,132 INFO coststat_import_all:calls=1,costs=1028098287,cost_per_call=1028098287,percents=100.0000%
[25282] 05-26 10:10:00,132 INFO coststat_import_range[1000000-]:calls=1,costs=1028098287,cost_per_call=1028098287,percents=100.0000%
[25282] 05-26 10:10:00,132 INFO slave_sync_total_commands_processed:0
[25282] 05-26 10:10:00,132 INFO slave_sync_instantaneous_ops_per_sec:0
[25282] 05-26 10:10:00,132 INFO total_commands_processed:1
[25282] 05-26 10:10:00,132 INFO instantaneous_ops_per_sec:0
[25282] 05-26 10:10:00,132 INFO total_connections_received:1
[25282] 05-26 10:10:00,132 INFO rejected_connections:0
[25282] 05-26 10:10:00,132 INFO ========================Period Statistics Dump End===========================`
``

**Comment by yinqiwen** [comment]
In the log, the `import` cost `1028098287us` to complete, but the `save` is not.   
And the dump file would be created in `$ARDB_HOME/backup`, u can find if it stop increasing the file size while u waiting the `save` complete.

**Comment by harutyundermenjyan** [comment]
yes, that it is. I mean the save command action in the redis-cli it doesn't returns any status after the save completion.

**Comment by yinqiwen** [comment]
is  this commit https://github.com/yinqiwen/ardb/commit/c7e3efe5e80bbf080dc2aff4d2608c58db157cf6 fixed the save hang issue?  
u can use `top` to see that if it's still in an infinite loop.

**Comment by harutyundermenjyan** [comment]
After the commit c7e3efe seems it can completely save the file but in the redis-cli it doesn't returns any status after the save completion.

**Comment by yinqiwen** [comment]
`save` command is not a background task, if it's not return, then it's not done at server.   

Any found in `top` command output for `ardb-server`?

**Comment by harutyundermenjyan** [comment]
PID USER      PR  NI    VIRT      RES      SHR S   %CPU    %MEM     TIME+      COMMAND
25282 root      20   0   2472532  0.988g   5692 R     100.1      0.8       207:44.36    ardb-server /etc/ardb.conf

**Comment by yinqiwen** [comment]
The cpu usage is 100%, it seems there is still an infinite loop issue.

**Comment by harutyundermenjyan** [comment]
yes you are noticed right

**Comment by yinqiwen** [comment]
can u use `gdb ./ardb-server <pid>` to attach running process to get the running stack trace?

**Comment by harutyundermenjyan** [comment]
I have performed gdb ./ardb-server <pid>
send the output ?

**Comment by yinqiwen** [comment]
`gdb ./ardb-server   25282`  in your env,  `25282` is the running ardb-server pid.

**Comment by harutyundermenjyan** [comment]
yes, I have already done
do I send the output ?

**Comment by yinqiwen** [comment]
use `bt` to see current thread's stack trace
use `info threads` to list all threads
use `thread <id>` to switch thread

**Comment by harutyundermenjyan** [comment]
OK done.
then?

**Comment by yinqiwen** [comment]
can u send the output here?

**Comment by harutyundermenjyan** [comment]
(gdb) info threads
  Id   Target Id         Frame 
  24   Thread 0x7f75417ff700 (LWP 25283) "rocksdb:bg0" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  23   Thread 0x7f7540ffe700 (LWP 25284) "rocksdb:bg1" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  22   Thread 0x7f75403ff700 (LWP 25285) "rocksdb:bg2" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  21   Thread 0x7f753fbfe700 (LWP 25286) "rocksdb:bg3" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  20   Thread 0x7f753edff700 (LWP 25287) "rocksdb:bg4" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  19   Thread 0x7f753e1ff700 (LWP 25288) "rocksdb:bg5" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  18   Thread 0x7f753d3ff700 (LWP 25289) "rocksdb:bg6" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  17   Thread 0x7f753c5ff700 (LWP 25290) "rocksdb:bg7" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  16   Thread 0x7f753b9ff700 (LWP 25291) "rocksdb:bg8" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  15   Thread 0x7f753adff700 (LWP 25292) "rocksdb:bg9" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  14   Thread 0x7f753a1ff700 (LWP 25293) "rocksdb:bg10" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  13   Thread 0x7f75395ff700 (LWP 25294) "rocksdb:bg11" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  12   Thread 0x7f75389ff700 (LWP 25295) "rocksdb:bg12" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  11   Thread 0x7f7537dff700 (LWP 25296) "rocksdb:bg13" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  10   Thread 0x7f75371ff700 (LWP 25297) "rocksdb:bg14" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  9    Thread 0x7f75365ff700 (LWP 25298) "rocksdb:bg15" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  8    Thread 0x7f752dbff700 (LWP 25299) "ardb-server" 0x00007f7541e63b13 in epoll_wait () at ../sysdeps/unix/syscall-template.S:81
  7    Thread 0x7f752cbff700 (LWP 25300) "ardb-server" 0x00007f7541e63b13 in epoll_wait () at ../sysdeps/unix/syscall-template.S:81
  6    Thread 0x7f752bdff700 (LWP 25301) "ardb-server" 0x00007f7541e63b13 in epoll_wait () at ../sysdeps/unix/syscall-template.S:81
  5    Thread 0x7f752a7ff700 (LWP 25310) "rocksdb:bg0" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  4    Thread 0x7f7529ffe700 (LWP 25311) "rocksdb:bg1" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  3    Thread 0x7f75293ff700 (LWP 25312) "rocksdb:bg2" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  2    Thread 0x7f75287ff700 (LWP 25313) "rocksdb:bg3" pthread_cond_wait@@GLIBC_2.3.2 () at ../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
* 1    Thread 0x7f75433ae980 (LWP 25282) "ardb-server" 0x00000000006771be in je_tcache_alloc_easy (tcache_success=<synthetic pointer>, tbin=0x7f754180c0c8)
    at include/jemalloc/internal/tcache.h:259
(gdb) thread 1
[Switching to thread 1 (Thread 0x7f75433ae980 (LWP 25282))]
#0  0x00000000006771be in je_tcache_alloc_easy (tcache_success=<synthetic pointer>, tbin=0x7f754180c0c8) at include/jemalloc/internal/tcache.h:259
259		if (unlikely(tbin->ncached == 0)) {
(gdb)

**Comment by yinqiwen** [comment]
seems the thread 1 is in infinite loop, just use `bt` to get the whole stack trace.

**Comment by harutyundermenjyan** [comment]
(gdb) bt
#0  0x00000000006771be in je_tcache_alloc_easy (tcache_success=<synthetic pointer>, tbin=0x7f754180c0c8) at include/jemalloc/internal/tcache.h:259
#1  je_tcache_alloc_small (size=<optimized out>, slow_path=false, zero=false, binind=5, tcache=0x7f754180c000, arena=0x0, tsd=<optimized out>)
    at include/jemalloc/internal/tcache.h:292
#2  je_arena_malloc (slow_path=false, tcache=0x7f754180c000, zero=false, ind=5, size=<optimized out>, arena=0x0, tsd=<optimized out>) at include/jemalloc/internal/arena.h:1276
#3  je_iallocztm (slow_path=false, arena=0x0, is_metadata=false, tcache=0x7f754180c000, zero=false, ind=5, size=<optimized out>, tsd=<optimized out>)
    at include/jemalloc/internal/jemalloc_internal.h:935
#4  je_imalloc (slow_path=false, ind=5, size=<optimized out>, tsd=<optimized out>) at include/jemalloc/internal/jemalloc_internal.h:954
#5  imalloc_body (slow_path=false, usize=<synthetic pointer>, tsd=<synthetic pointer>, size=<optimized out>) at src/jemalloc.c:1467
#6  malloc (size=<optimized out>) at src/jemalloc.c:1504
#7  0x00007f75426a8dad in operator new(unsigned long) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#8  0x00007f75426a8ea9 in operator new[](unsigned long) () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#9  0x000000000063b1b8 in TrimAppend (non_shared_len=75, 
    non_shared_data=0x7f74fea00003 "\036similar_images:201581489003202\n\002\002\316\357Ϛ\336\364\246\332\177\004\026photos:175898142000201\001", shared_len=0, 
    this=0x7f7536625708) at ./db/dbformat.h:302
#10 rocksdb::BlockIter::ParseNextKey (this=0x7f75366256c0) at table/block.cc:156
#11 0x00000000005cd346 in SeekToFirst (this=0x7f754199bdd0) at ./table/iterator_wrapper.h:58
#12 rocksdb::(anonymous namespace)::TwoLevelIterator::SeekToFirst (this=0x7f754199bda0) at table/two_level_iterator.cc:107
#13 0x00000000005cd216 in SeekToFirst (this=0x7f754195c220) at ./table/iterator_wrapper.h:58
#14 rocksdb::(anonymous namespace)::TwoLevelIterator::SkipEmptyDataBlocksForward (this=0x7f754195c1d0) at table/two_level_iterator.cc:149
#15 0x00000000005c5337 in Seek (k=..., this=0x7f754195bfb0) at ./table/iterator_wrapper.h:57
#16 rocksdb::MergingIterator::Seek (this=0x7f754195bf50, target=...) at table/merger.cc:105
#17 0x000000000057292f in rocksdb::DBIter::Seek (this=0x7f754195be38, target=...) at db/db_iter.cc:677
#18 0x000000000053058c in ardb::RocksDBIterator::Jump (this=0x7f75418e3400, next=...) at db/rocksdb/rocksdb_engine.cpp:1304
#19 0x0000000000523507 in ardb::Snapshot::RedisSave (this=0x7f7540458b40) at repl/rdb.cpp:2165
#20 0x0000000000523c95 in ardb::Snapshot::DoSave (this=this@entry=0x7f7540458b40) at repl/rdb.cpp:1788
#21 0x0000000000523cf0 in ardb::Snapshot::Save (this=this@entry=0x7f7540458b40, type=type@entry=ardb::REDIS_DUMP, file="/var/lib/ardb/backup/redis-snapshot.1464243198", 
    cb=cb@entry=0x4cbd10 <ardb::RDBSaveLoadRoutine(ardb::SnapshotState, ardb::Snapshot*, void*)>, data=data@entry=0x7f75419b7680) at repl/rdb.cpp:1837
#22 0x00000000005271bd in ardb::SnapshotManager::NewSnapshot (this=0x988700 <ardb::g_snapshot_manager_instance>, type=type@entry=ardb::REDIS_DUMP, bgsave=bgsave@entry=false, 
    cb=cb@entry=0x4cbd10 <ardb::RDBSaveLoadRoutine(ardb::SnapshotState, ardb::Snapshot*, void*)>, data=data@entry=0x7f75419b7680) at repl/rdb.cpp:2459
#23 0x00000000004cc034 in ardb::Ardb::Save (this=<optimized out>, ctx=..., cmd=...) at command/server.cpp:71
#24 0x0000000000505e80 in ardb::Ardb::DoCall (this=this@entry=0x7ffeaeea7840, ctx=..., setting=..., args=...) at db/db.cpp:1301
#25 0x0000000000509a2f in ardb::Ardb::Call (this=0x7ffeaeea7840, ctx=..., args=...) at db/db.cpp:1476
#26 0x000000000045d7b9 in ardb::RedisRequestHandler::MessageReceived (this=0x7f753c88c180, ctx=..., e=...) at network.cpp:87
#27 0x000000000045b193 in HandleStreamEvent (e=..., ctx=..., this=<optimized out>) at ./common/channel/channel_upstream_handler.hpp:133
#28 ardb::ChannelPipeline::SendUpstream<ardb::codec::RedisCommandFrame> (ctx=<optimized out>, e=..., this=<optimized out>) at ./common/channel/all_includes.hpp:89
#29 0x000000000045f2be in SendUpstream<ardb::MessageEvent<ardb::codec::RedisCommandFrame> > (this=<optimized out>, this=<optimized out>, e=...)
    at ./common/channel/all_includes.hpp:164
#30 fire_message_received<ardb::codec::RedisCommandFrame> (destructor=0x0, message=0x7ffeaeea7460, ctx=...) at ./common/channel/channel_helper.hpp:91
#31 ardb::codec::StackFrameDecoder<ardb::codec::RedisCommandFrame>::CallDecode (this=0x7f7541959740, context=..., channel=0x7f75419b7f40, cumulation=...)
    at ./common/channel/codec/stack_frame_decoder.hpp:102
#32 0x000000000045f5ff in ardb::codec::StackFrameDecoder<ardb::codec::RedisCommandFrame>::MessageReceived (this=0x7f7541959740, ctx=..., e=...)
    at ./common/channel/codec/stack_frame_decoder.hpp:157
#33 0x0000000000461e33 in HandleStreamEvent (e=..., ctx=..., this=<optimized out>) at ./common/channel/channel_upstream_handler.hpp:133
#34 ardb::ChannelPipeline::SendUpstream<ardb::Buffer> (ctx=<optimized out>, e=..., this=<optimized out>) at ./common/channel/all_includes.hpp:89
#35 0x0000000000463340 in SendUpstream<ardb::MessageEvent<ardb::Buffer> > (event=..., this=0x7f75419b7f78) at ./common/channel/all_includes.hpp:128
---Type <return> to continue, or q <return> to quit--- 
#36 fire_message_received<ardb::Buffer> (destructor=0x0, message=0x7f75419b7fa8, channel=0x7f75419b7f40) at ./common/channel/channel_helper.hpp:83
#37 ardb::Channel::OnRead (this=0x7f75419b7f40) at common/channel/channel.cpp:550
#38 0x0000000000461c5d in ardb::Channel::IOEventCallback (eventLoop=<optimized out>, fd=<optimized out>, clientData=0x7f75419b7f40, mask=1) at common/channel/channel.cpp:53
#39 0x000000000049e1a0 in aeProcessEvents (eventLoop=eventLoop@entry=0x7f7541823078, flags=flags@entry=3) at common/channel/redis/ae.c:429
#40 0x000000000049e48b in aeMain (eventLoop=0x7f7541823078) at common/channel/redis/ae.c:485
#41 0x000000000045bae8 in ardb::Server::Start (this=this@entry=0x7ffeaeea7820) at network.cpp:296
#42 0x000000000040abac in main (argc=<optimized out>, argv=0x7ffeaeea7ce8) at main.cpp:92
(gdb)

**Comment by yinqiwen** [comment]
this seems not an  infinite loop, if it's an  infinite loop, the stack trace would hundreds of frames.  

Is your env has a limit memory? seems that blocked in jemalloc to allocate memory.

**Comment by harutyundermenjyan** [comment]
root@supermicro1:~# ulimit -a
core file size          (blocks, -c) 0
data seg size           (kbytes, -d) unlimited
scheduling priority             (-e) 0
file size               (blocks, -f) unlimited
pending signals                 (-i) 514822
max locked memory       (kbytes, -l) 64
max memory size         (kbytes, -m) unlimited
open files                      (-n) 1024
pipe size            (512 bytes, -p) 8
POSIX message queues     (bytes, -q) 819200
real-time priority              (-r) 0
stack size              (kbytes, -s) 8192
cpu time               (seconds, -t) unlimited
max user processes              (-u) 514822
virtual memory          (kbytes, -v) unlimited
file locks                      (-x) unlimited

**Comment by harutyundermenjyan** [comment]
seems it ardb code specific something

**Comment by harutyundermenjyan** [comment]
Is there any issue?

**Comment by yinqiwen** [comment]
Can u update source to rebuild  and test again? Just add some log for saving snapshot.

**Comment by yinqiwen** [comment]
And the log level should be set to DEBUG

**Comment by harutyundermenjyan** [comment]
the problem still exext

**Comment by harutyundermenjyan** [comment]
the problem still exist
here is the output of DEBUG it is infinitely adding this line
[10844] 05-26 20:10:37,081 DEBUG Save key/value with type:11 & key:similar_images:199772818000202
[10844] 05-26 20:10:37,081 DEBUG Save key/value with type:11 & key:similar_images:199772818000202
[10844] 05-26 20:10:37,081 DEBUG Save key/value with type:11 & key:similar_images:199772818000202
[10844] 05-26 20:10:37,081 DEBUG Save key/value with type:11 & key:similar_images:199772818000202
[10844] 05-26 20:10:37,081 DEBUG Save key/value with type:11 & key:similar_images:199772818000202
[10844] 05-26 20:10:37,081 DEBUG Save key/value with type:11 & key:similar_images:199772818000202
[10844] 05-26 20:10:37,081 DEBUG Save key/value with type:11 & key:similar_images:199772818000202
[10844] 05-26 20:10:37,081 DEBUG Save key/value with type:11 & key:similar_images:199772818000202
[10844] 05-26 20:10:37,081 DEBUG Save key/value with type:11 & key:similar_images:199772818000202
[10844] 05-26 20:10:37,081 DEBUG Save key/value with type:11 & key:similar_images:199772818000202
[10844] 05-26 20:10:37,081 DEBUG Save key/value with type:11 & key:similar_images:199772818000202
[10844] 05-26 20:10:37,081 DEBUG Save key/value with type:11 & key:similar_images:199772818000202
[10844] 05-26 20:10:37,081 DEBUG Save key/value with type:11 & key:similar_images:199772818000202
[10844] 05-26 20:10:37,081 DEBUG Save key/value with type:11 & key:similar_images:199772818000202
[10844] 05-26 20:10:37,081 DEBUG Save key/value with type:11 & key:similar_images:199772818000202

**Comment by harutyundermenjyan** [comment]
seem it hangs up in this stage and infinitely remains in save mode

**Comment by yinqiwen** [comment]
Can u comment lines https://github.com/yinqiwen/ardb/blob/0.9/src/repl/rdb.cpp#L2170-L2172   & rebuild to test again? It seems  a issue related rocksdb, which cause that the code at line https://github.com/yinqiwen/ardb/blob/0.9/src/repl/rdb.cpp#L2171 never work.

And u can build ardb-server with leveldb to test if it's  a rocksdb relates issue.

**Comment by yinqiwen** [comment]
no need to do that, i just recreate this issue.

**Comment by yinqiwen** [comment]
it's a rocksdb related issue, fixed by latest commit.

**Comment by harutyundermenjyan** [comment]
Dear Yinqiwen
The issue really has been solved.
I'm now going to test other dbs engines especially WiredTiger.
Thanks a lot!

**Comment by harutyundermenjyan** [comment]
127.0.0.1:16379> save
OK
(593.33s)
127.0.0.1:16379> 
: )

**Comment by harutyundermenjyan** [comment]
can I use for future the saved file by ardb redis-snapshot.1464335990 for redis as dump.rdb ?
for example:
mv /var/lib/ardb/backup/redis-snapshot.1464335990  /var/lib/redis/dump.rdb 
and then start the redis ?
Is the saved format by ardb redis dump format compatible?

**Comment by yinqiwen** [comment]
yes，it is redis fornat. Use  save2 would  generate ardb format snapshot file

**Comment by harutyundermenjyan** [comment]
Then it seem we have another issue that is when I am trying to start redis with dump file which previously saved by ardb save command the redis crashes.
here my steps:
1. mv /var/lib/ardb/backup/redis-snapshot.1464335990 /var/lib/redis/dump.rdb
2. service redis-server start   here the redis crashes
Please see the redis log file output below:
[24753] 27 May 14:13:46.679 # --- CLIENT LIST OUTPUT
[24753] 27 May 14:13:46.679 # 
[24753] 27 May 14:13:46.679 # --- REGISTERS
[24753] 27 May 14:13:46.679 # 
RAX:0000000000000000 RBX:0000000000473684
RCX:0000000000000005 RDX:0000000000000006
RDI:00007fed03806000 RSI:000000000000003c
RBP:0000000000474da5 RSP:00007fff65aa5240
R8 :00007fed038427e0 R9 :00007fed04e30780
R10:00007fed044bd6a0 R11:0000000000000206
R12:0000000000000405 R13:0000000000000000
R14:0000000000000031 R15:00007fed038eb300
RIP:000000000044202a EFL:0000000000010206
CSGSFS:0000000000000033
[24753] 27 May 14:13:46.679 # (00007fff65aa524f) -> 0000000000000031
[24753] 27 May 14:13:46.679 # (00007fff65aa524e) -> 0000000000000000
[24753] 27 May 14:13:46.679 # (00007fff65aa524d) -> 00007fed03842480
[24753] 27 May 14:13:46.679 # (00007fff65aa524c) -> 00007fed03815980
[24753] 27 May 14:13:46.679 # (00007fff65aa524b) -> ffffffffffffffff
[24753] 27 May 14:13:46.679 # (00007fff65aa524a) -> 0000000000000003
[24753] 27 May 14:13:46.679 # (00007fff65aa5249) -> 000000000042b50a
[24753] 27 May 14:13:46.679 # (00007fff65aa5248) -> 0000000000000039
[24753] 27 May 14:13:46.679 # (00007fff65aa5247) -> 00007fff65aa5300
[24753] 27 May 14:13:46.679 # (00007fff65aa5246) -> 0000000000000039
[24753] 27 May 14:13:46.679 # (00007fff65aa5245) -> 00000000004247c8
[24753] 27 May 14:13:46.679 # (00007fff65aa5244) -> 00007fed03810658
[24753] 27 May 14:13:46.679 # (00007fff65aa5243) -> 000000000042ceea
[24753] 27 May 14:13:46.679 # (00007fff65aa5242) -> 00007fed03842480
[24753] 27 May 14:13:46.679 # (00007fff65aa5241) -> 0000000000000031
[24753] 27 May 14:13:46.679 # (00007fff65aa5240) -> 00007fff65aa5300
[24753] 27 May 14:13:46.679 # --- FAST MEMORY TEST
[24753] 27 May 14:13:46.679 # Bio thread for job type #0 terminated
[24753] 27 May 14:13:46.679 # Bio thread for job type #1 terminated
[24753] 27 May 14:13:47.712 # Fast memory test PASSED, however your memory can still be broken. Please run a memory test for several hours if possible.
[24753] 27 May 14:13:47.712 # 
=== REDIS BUG REPORT END. Make sure to include from START to END. ===

       Please report the crash opening an issue on github:

           http://github.com/antirez/redis/issues

  Suspect RAM error? Use redis-server --test-memory to veryfy it.

**Comment by harutyundermenjyan** [comment]
in case of there is a lot of free mem on the server

**Comment by harutyundermenjyan** [comment]
I have opened another issue for this please have a look to issue #248
